### PR TITLE
 coord,persist: teach coord about persisted streams that are involved in sources

### PR DESCRIPTION
**Pull request by aljoscha**
Before, the coordinator would only know the "prefix" that was used to
form the names of persistent streams that are involved in, say, the
persistent Kafka source.

Now, we keep a mapping from a _logical name_ to the physical stream name
in the coordinator, for all persisted streams that are involved in a
source.

This also changes what persist information is sent to workers. Before,
we would send a single persist name. Now, we send a new
`SourcePersistDesc`, which contains the mapping of streams. In future
commits, we will add more information to this.

We need to do this to allow the coordinator to:
 - determine a common seal timestamp when restarting
 - negotiate/figure out if a requested as_of matches the compaction
   frontier/since of a persisted source
 - compact persisted streams when necessary

We will do these additional changes in follow-up commits.

### Tips for reviewer

The contentious change is how I manually serialize (and deserialize) a mapping of names to the one `persist_name` field that we store in the catalog. I'm doing this in order to not break that, but could be convinced to do a more elaborate thing.

Also, there are at least two `TODO:...`s in the changes that are questions, please take a look at them.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Comment by aljoscha** [comment]
I _think_ this breaks the builtin table, because we now render the `persist_name` differently, that is we print the mappings instead of a single name. I'll look into fixing that.

**Comment by aljoscha** [comment]
Also: Happy to change all the mentions of _persisted stream_ to _persisted collection_.

File: src/coord/src/persistcfg.rs [hunk 1]
```diff
@@ -197,14 +197,18 @@ impl PersisterWithConfig {
         }
     }
 
-    pub fn details(&self, id: GlobalId, pretty: &str) -> Result<Option<PersistDetails>, Error> {
+    pub fn details(
```
Comment (danhhz) [inline src/coord/src/persistcfg.rs:201]: let's call this table_details now (ditto for table_details_from_name). maybe for the stuff in catalog.rs too?

**Review by danhhz — COMMENTED** [review]
hmmm, I don't love this, but I don't have enough background to have a specific counter-proposal. can we schedule a time to chat and whiteboard some other approaches?

**Comment by aljoscha** [comment]
Closing in favour of #9449

**Pull request closed without merging**